File: optimizer/taxing/tax_rooms.py
"""
Detects room collisions along the rasp's all dates path and taxes them.
"""
def tax_rrule_in_rooms(state, room_id, rasp):
    all_dates      = state.rasp_rrules[rasp.id]["all_dates"]
    room_occupied  = state.mutable_constraints.rooms_occupied[room_id]

    cnt = 0
    for week, day, hour in all_dates:
        room_occupied[week, day, hour] += 1
        cnt += room_occupied[week, day, hour] if room_occupied[week, day, hour] > 1 else 0
    if cnt:
        punish = -cnt*30
        update_grade_rooms(state, punish, plus=True)


"""
Detects room collisions along the rasp's all dates path and untaxes them.
Used to undo the previous tax.
"""
def untax_rrule_in_rooms(state, room_id, rasp):
    all_dates      = state.rasp_rrules[rasp.id]["all_dates"]
    room_occupied  = state.mutable_constraints.rooms_occupied[room_id]

    cnt = 0
    for week, day, hour in all_dates:
        cnt += room_occupied[week, day, hour] if room_occupied[week, day, hour] > 1 else 0
        room_occupied[week, day, hour] -= 1
    if cnt:
        punish = -cnt*30
        update_grade_rooms(state, punish, plus=False)


"""
Updates the grade with calculated "punish" score.
"""
def update_grade_rooms(state, punish, plus=True):
    grade = state.grade
    if plus:
        grade["roomScore"] += punish
        grade["totalScore"] += punish
    else:
        grade["roomScore"] -= punish
        grade["totalScore"] -= punish

File: cpp_optimizer/taxing/tax_rooms.cpp
#include "tax_rooms.h"
#include <iostream>
#include <vector>

void update_grade_rooms(State& state, int punish, bool plus) {
    if(plus) {
        state.grade.roomScore  += punish;
        state.grade.totalScore += punish;
    }
    else {
        state.grade.roomScore  -= punish;
        state.grade.totalScore -= punish;
    }
}

void tax_rrule_in_rooms(State& state, id_room room_id, Rasp& rasp) {
    array3D room_occupied = state.mutable_constraints.rooms_occupied[room_id];
    std::vector<triplet>& all_dates = state.rasp_rrules[rasp.id].all_dates;

    int cnt = 0;
    for(triplet date : all_dates) {
        int week = std::get<0>(date);
        int day  = std::get<1>(date);
        int hour = std::get<2>(date);
        room_occupied[week][day][hour] +=1;
        int collisions = room_occupied[week][day][hour];
        cnt = (collisions > 1) ? (cnt + collisions) : cnt;
    }

    if(cnt) {
        int punish = -cnt * 30;
        update_grade_rooms(state, punish, true);
    }
}

void untax_rrule_in_rooms(State& state, id_room room_id, Rasp& rasp) {
    array3D room_occupied = state.mutable_constraints.rooms_occupied[room_id];
    std::vector<triplet>& all_dates = state.rasp_rrules[rasp.id].all_dates;

    int cnt = 0;
    for(triplet date : all_dates) {
        int week = std::get<0>(date);
        int day  = std::get<1>(date);
        int hour = std::get<2>(date);
        int collisions = room_occupied[week][day][hour];
        cnt = (collisions > 1) ? (cnt + collisions) : cnt;
        room_occupied[week][day][hour] -=1;
    }

    if(cnt) {
        int punish = -cnt * 30;
        update_grade_rooms(state, punish, false);
    }
}
